File: sliding-window/max-sum-subarray/max-sum-subarray.py
"""Sliding Window: Maximum sum of a subarray of size k"""


def max_sum_subarray(nums: list[int], k: int) -> int:
    window_sum = sum(nums[:k])
    max_sum = window_sum

    for i in range(k, len(nums)):
        window_sum += nums[i] - nums[i - k]
        max_sum = max(max_sum, window_sum)

    return max_sum


if __name__ == "__main__":
    nums = [2, 1, 5, 1, 3, 2]
    k = 3

    print(f"Max sum: {max_sum_subarray(nums, k)}")

File: sliding-window/max-sum-subarray/max-sum-subarray.cpp
// Sliding Window: Maximum sum of a subarray of size k
#include <iostream>
#include <vector>
using namespace std;

int maxSumSubarray(const vector<int>& nums, int k) {
    int windowSum = 0;
    for (int i = 0; i < k; i++) windowSum += nums[i];

    int maxSum = windowSum;
    for (int i = k; i < (int)nums.size(); i++) {
        windowSum += nums[i] - nums[i - k];
        maxSum = max(maxSum, windowSum);
    }
    return maxSum;
}

int main() {
    vector<int> nums = {2, 1, 5, 1, 3, 2};
    int k = 3;

    cout << "Max sum: " << maxSumSubarray(nums, k) << endl;

    return 0;
}
